### PR TITLE
serialise_executable: Name the data file deterministically

### LINKED ISSUE #1062: custom_target exe-wrapper data filename always changes on regen
I have custom target with LONG command line (around 500 files at once)
Simplified project:

[classpath.zip](https://www.dropbox.com/s/nsh2y46e6opqrw1/classpath.zip)
```
project('avian', ['cpp'])

subdir('classpath')

javac = find_program('javac')

classpath_classes = custom_target('classpath_classes',
  input: classpath_java_sources,
  output: 'classpath_classes',
  capture: true,
  command: [javac,
    '-encoding', 'UTF-8',
    '-source', '1.6', # JAVA_VERSION,
    '-target', '1.6', # JAVA_VERSION,
    '-d', meson.current_build_dir() + '/classpath',
    '-bootclasspath', meson.current_build_dir() + '/classpath',
    '@INPUT@'
  ],
)
```

Commands:
```
$ mkdir build
$ meson build
$ ninja -C build -d explain classpath_classes
```
Build success.
Next time:
```
$ ninja -C build -d explain classpath_classes
ninja: Entering directory `build'
ninja: no work to do.
```
But:
```
$ touch meson.build
$ ninja -C build -d explain classpath_classes
ninja: Entering directory `build'
ninja explain: output build.ninja older than most recent input ../meson.build (501180038 vs 501180064)
[0/1] Regenerating build files
The Meson build system
Version: 0.36.0
Source dir: F:\avian\classpath
Build dir: F:\avian\classpath\build
Build type: native build
Project name: avian
Native cpp compiler: c++ (gcc 6.2.0)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Program javac found: YES (C:\Program Files\Java\jdk1.8.0_111\bin\javac.EXE)
Build targets in project: 1
ninja explain: command line changed for classpath_classes
```

Where is command line changed?

### PR DESCRIPTION
**Pull request by nirbheek**
Includes a unit test for the same.

Closes #1062

### CODE REVIEW

File: mesonbuild/backend/backends.py [hunk 1]
```diff
@@ -197,15 +197,20 @@ def flatten_object_list(self, target, proj_dir_to_build_root=''):
...
                             build.BuildTarget, build.CustomTarget)):
             basename = exe.name
         else:
             basename = os.path.basename(exe)
-        scratch_file = 'meson_exe_{0}_{1}.dat'.format(basename,
-                                                      str(uuid.uuid4())[:8])
+        # Take a digest of the cmd args, env, workdir, and capture. This avoids
+        # collisions and also makes the name deterministic over regenerations
+        # which avoids a rebuild by Ninja because the cmdline stays the same.
+        data = bytes(str(env) + str(cmd_args) + str(workdir) + str(capture),
```
Comment (QuLogic) [inline mesonbuild/backend/backends.py]: Isn't `env` a dictionary? How are you ensuring the result is sorted the same way?

Comment (nirbheek) [inline mesonbuild/backend/backends.py]: The order in which key/value pairs are popped from a dict is unspecified but is stable. For all known versions of Python, if you insert items in two dictionaries in a specific order, the items will pop in a (different) specific order. You can check this yourself, and the unit test also checks for this.

In fact, starting from 3.6, it is specified to always [return in the same order](https://mail.python.org/pipermail/python-dev/2016-September/146327.html) in which the items were inserted, so we will always be able to rely on this.

Comment (TingPing) [inline mesonbuild/backend/backends.py]: @nirbheek Unless we only want to support CPython the fact dicts are sorted in 3.6+ is an implementation detail and you are still supposed to use `OrderedDict`

Comment (nirbheek) [inline mesonbuild/backend/backends.py]: That's why we have the unit test, but if you insist, I will change the implementation to be sorted.

Comment (TingPing) [inline mesonbuild/backend/backends.py]: I don't think we test alternate implementations?

Comment (nirbheek) [inline mesonbuild/backend/backends.py]: No, but if someone starts to test with another implementation, we will know. Anyway, this is already fixed.

**Comment by jpakkane** [comment]
:crossed_swords: conflicts.

**Comment by nirbheek** [comment]
Fixed!

**Pull request merged**